server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CachedDumbBlobStore.java

```diff
@@ -0,0 +1,138 @@
+import org.apache.james.blob.api.ObjectStoreIOException;
+import org.apache.james.blob.cassandra.CassandraBucketDAO;
+import org.reactivestreams.Publisher;
+
+import com.google.common.base.Preconditions;
+import com.google.common.io.ByteSource;
+
+import reactor.core.publisher.Mono;
+
+public class CachedDumbBlobStore implements DumbBlobStore {
+
+    public static final String DEFAULT_BUCKET = "cassandraDefault";
+
+    private final DumbBlobStoreCache cache;
+    private final DumbBlobStore backend;
+    private final Integer sizeThresholdInBytes;
+    private final BucketName defaultBucket;
+    private final CassandraBucketDAO bucketDAO;
+
+    @Inject
+    public CachedDumbBlobStore(DumbBlobStoreCache cache, DumbBlobStore backend,
+                               CassandraCacheConfiguration cacheConfiguration,
+                               CassandraBucketDAO bucketDAO,
+                               @Named(DEFAULT_BUCKET) BucketName defaultBucket) {
+        this.cache = cache;
+        this.backend = backend;
+        this.sizeThresholdInBytes = cacheConfiguration.getSizeThresholdInBytes();
+        this.defaultBucket = defaultBucket;
+        this.bucketDAO = bucketDAO;
+    }
+
+    @Override
+    public InputStream read(BucketName bucketName, BlobId blobId) throws ObjectStoreIOException, ObjectNotFoundException {
+        Preconditions.checkNotNull(bucketName, "bucketName should not be null");
+
+        return Mono.just(bucketName)
+            .filter(defaultBucket::equals)
+            .flatMap(bucker ->
+                Mono.from(cache.read(blobId))
+                    .<InputStream>flatMap(bytes -> Mono.fromCallable(() -> new ByteArrayInputStream(bytes)))
+                    .switchIfEmpty(Mono.fromCallable(() -> backend.read(bucketName, blobId)))
+            )
+            .switchIfEmpty(Mono.fromCallable(() -> backend.read(bucketName, blobId)))
+            .blockOptional()
+            .orElseThrow(() -> new ObjectNotFoundException(String.format("Could not retrieve blob metadata for %s", blobId)));
+    }
+
+    @Override
+    public Publisher<byte[]> readBytes(BucketName bucketName, BlobId blobId) {
+        return Mono.from(cache.read(blobId))
+            .switchIfEmpty(Mono.from(backend.readBytes(bucketName, blobId)));
+    }
+
+    @Override
+    public Publisher<Void> save(BucketName bucketName, BlobId blobId, byte[] data) {
+        return Mono.from(backend.save(bucketName, blobId, data))
+            .then(Mono.just(bucketName)
+                .filter(bucket -> bucket.equals(defaultBucket) && data.length <= sizeThresholdInBytes)
+                .flatMap(ignored -> Mono.from(cache.cache(blobId, data))));
+    }
+
+    @Override
+    public Publisher<Void> save(BucketName bucketName, BlobId blobId, InputStream inputStream) {
+        return Mono.from(backend.save(bucketName, blobId, inputStream))
+            .then(Mono.just(inputStream)
+                .flatMap(stream -> handleCache(bucketName, blobId, stream)));
```

Comment:
This is not optimal as we end up reading the backend to populate the cache.

We can do:

```
if (bucketName.equals(defaultBucket) {
    return backend.save(bucketName, blobId, inputStream);
}

BufferedInputStream buf = new BufferedInputStream(inputStream);
return isItABigStream(buf) // returns a mono of boolean
    .flatMap(isBigStream -> {
        if (isBigStream) {
             return Mono.from(backend.save(bucketName, blobId, buf));
        }
        byte[] content = IOUtilstoByteArray(buf); // legitimate as the stream is small
        return Mono.from(backend.save(bucketName, blobId, content))
             .then(cache.save(blobId, content);
    });
```

Reply:
I know that we already use `BufferedInputStream` that way but it's wrong: you don't have any guarantee about what is buffered and the implementation can change.
What you need is a `PushbackInputStream`

Reply:
Ok